### PR TITLE
ISPN-12594 Exception in cache streams with DENY_READ_WRITES

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-12594

Use UnicastProcessor instead of PublishProcessor,
as PublishProcessor starts emitting items too early
(before its subscribers call request()).

### CODE REVIEW

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorTest.java [hunk 1]
```diff
@@ -296,7 +296,7 @@ public void testNodeLeavesWhileIteratingOverContainerCausingRehashToLoseValues()
       final BlockingQueue<Map.Entry<Object, String>> returnQueue = new LinkedBlockingQueue<>();
       Future<Void> future = fork(() -> {
          // Put batch size to a lower number just to make sure it doesn't retrieve them all in 1 go
-         Iterator<Map.Entry<Object, String>> iter = cache2.entrySet().stream().distributedBatchSize(2).iterator();
+         Iterator<Map.Entry<Object, String>> iter = cache2.entrySet().stream().distributedBatchSize(cache2.size() - 2).iterator();
```
Comment (wburns) [inline core/src/test/java/org/infinispan/stream/DistributedStreamIteratorTest.java:299]: Is this change required?

Comment (danberindei) [reply core/src/test/java/org/infinispan/stream/DistributedStreamIteratorTest.java:299]: No, it was an attempt to reproduce before I had the new test.

**Comment by wburns** [comment]
The change itself seems fine, but want to dig a bit deeper into it. Cause we are also using the PublishProcessor at https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java#L151 and want to understand if it is safe or not.

**Comment by wburns** [comment]
Talking further we should probably change https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java#L151 to use the UnicastProcessor as well, since a user could technically use a backpressure required operation in the finalizer.

**Comment by danberindei** [comment]
Updated @wburns

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !
